src/Microsoft.SqlTools.Authentication/Authenticator.cs

```diff
@@ -0,0 +1,242 @@
+                clientApplicationInstance = CreatePublicClientAppInstance(authority, audience);
+                s_pcaMap.TryAdd(authorityUrl, clientApplicationInstance);
+            }
+            return clientApplicationInstance;
+        }
+
+        private IPublicClientApplication CreatePublicClientAppInstance(string authority, string audience) =>
+            PublicClientApplicationBuilder.Create(ApplicationClientId)
+                .WithAuthority(authority, audience)
+                .WithClientName(ApplicationName)
+                .WithRedirectUri(RedirectUri)
+                .WithClientId(ApplicationClientId)
+                .WithLogging((logLevel, message, pii) =>
+                {
+                    switch (logLevel)
+                    {
+                        case LogLevel.Error:
+                            if (!pii) Utility.Logger.Error(message);
+                            break;
+                        case LogLevel.Warning:
+                            if (!pii) Utility.Logger.Warning(message);
+                            break;
+                        case LogLevel.Info:
+                            if (!pii) Utility.Logger.Information(message);
+                            break;
+                        case LogLevel.Verbose:
+                            if (!pii) Utility.Logger.Verbose(message);
+                            break;
+                        case LogLevel.Always:
+                            if (!pii) Utility.Logger.Critical(message);
+                            break;
+                    }
+                })
+                .Build();
+
+        private async Task<AuthenticationResult> AcquireTokenInteractiveAsync(IPublicClientApplication publicClientApplication, string[] scopes, Guid connectionId, string username, AuthenticationMethod authenticationMethod, CancellationToken cancellationToken)
+        {
+            try
+            {
+                CancellationTokenSource ctsInteractive = new CancellationTokenSource();
+                /*
+                 * On .NET Core, MSAL will start the system browser as a separate process. MSAL does not have control over this browser,
+                 * but once the user finishes authentication, the web page is redirected in such a way that MSAL can intercept the Uri.
+                 * MSAL cannot detect if the user navigates away or simply closes the browser. Apps using this technique are encouraged
+                 * to define a timeout (via CancellationToken). We recommend a timeout of at least a few minutes, to take into account
+                 * cases where the user is prompted to change password or perform 2FA.
+                 *
+                 * https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/System-Browser-on-.Net-Core#system-browser-experience
+                 */
+                ctsInteractive.CancelAfter(180000);
+
+                return await publicClientApplication.AcquireTokenInteractive(scopes)
```

Comment:
IIUC, this will cause a browser to be launched from STS?  Currently STS is designed to explicitly be a non-UI process, to support scenarios such as remote extension host or webmode.  We haven't made a decision that we won't support those scenarios in the future and having STS directly interact UI would not work in those scenarios.

Reply:
FWIW, thought what we were planning to support was ADS gettting token through the browser and then providing that token to SqlClient.  Not having STS show the browser itself.

Reply:
I'll see what best we can do, this should ideally never ever be hit.
Will throw error for now and trigger authentication in ADS when exception received.

Reply:
OK, that makes sense and I think I recall you mentioning this case in the design discussion, IIRC.  ADS doesn't currently ship in remote/web scenarios (though we may in the future if needed) but VS Code does, so this wouldn't work in those scenarios.  If there is a way to send the error to the UI client to launch the browser that would be a more maintainable/flexible design IMO.

Reply:
I think we can use `SecurityTokenRequest` to request token from ADS' registered Azure provider (Azure Core) asynchronously and be able to await on it, as done in `AzureActiveDirectoryAuthenticationCallback` in `ConnectionService`. However, the request/response format need a bit of rework to be able to pass necessary information and get back response with expiry date. I can use a `delegate` in `ConnectionService` that can make this request and send it to the `AuthenticationProvider`, to be able to attach Hosting service.

That should work 🤔

Reply:
@kburtram I've added design changes for this, please take a look.
Meanwhile I'm trying to run some tests by forcing this flow - in normal cases/cache resets I can't get this triggered yet.

Reply:
I tested this, and you were right, we do hit a deadlock with awaited call - I'm changing back to throwing exceptions so we can catch them when needed. At the moment, our design ensures user account is added and not stale before we even connect/reconnect, and MSAL token cache is reliable, so silent authentication should work every-time we hit STS including cases like OE expand/refresh/dashboard calls where token needs to be refreshed. This new class implementation will be able to always fetch token silently from MSAL cache.

If in any case exception is thrown, it will provide specific information so either we will be able to address the exception by catching and handling it or user will be prompted to take necessary action.